File: java/VmSimulation/src/parser/Parser.java
package parser;

import ast.*;
import lexer.Lexer;
import token.Token;
import token.TokenType;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Supplier;

public class Parser {
    private Token currentToken;
    private Token peekToken;
    private Lexer lexer;
    private String source;

    private Map<TokenType, Supplier<AstNode>> prefixParseFunctions = new HashMap<>();
    private Map<TokenType, Function<AstNode, AstNode>> infixParseFunctions = new HashMap<>();

    private Map<TokenType, PrecedenceTypeEnum> precedences = new HashMap<>();

    public Parser () {
        lexer = new Lexer();
        prefixParseFunctions.put(TokenType.INT, this::parseIntegerLiteral);
        infixParseFunctions.put(TokenType.PLUS, this::parseInfixExpression);
        infixParseFunctions.put(TokenType.MINUS, this::parseInfixExpression);
        infixParseFunctions.put(TokenType.DIVIDE, this::parseInfixExpression);
        infixParseFunctions.put(TokenType.MULTIPLY, this::parseInfixExpression);

        precedences.put(TokenType.PLUS, PrecedenceTypeEnum.SUM);
        precedences.put(TokenType.MINUS, PrecedenceTypeEnum.SUM);
        precedences.put(TokenType.DIVIDE, PrecedenceTypeEnum.PRODUCT);
        precedences.put(TokenType.MULTIPLY, PrecedenceTypeEnum.PRODUCT);
    }

    public void buildProgramStatements(AstProgram program) {
        while (currentToken.getType() != TokenType.EOF) {
            //System.out.println(currentToken);
            AstNode node = parseStatement();
            //System.out.println(node);
            if (node != null) {
                program.statements.add(node);
            }
            nextToken();
        }
    }

    private AstNode parseStatement() {
        return parseExpressionStatement();
    }

    private AstNode parseExpressionStatement() {
        AstExpressionStatement stmt = new AstExpressionStatement(currentToken);
        stmt.setExpression(parseExpression(PrecedenceTypeEnum.LOWEST));
        if (peekTokenIs(TokenType.SEMICOLON)) {
            nextToken();
        }
        return stmt;
    }

    private AstNode parseExpression(PrecedenceTypeEnum pte) {
        if (!prefixParseFunctions.containsKey(currentToken.getType())) {
            return null;
        }
        Supplier<AstNode> prefix = prefixParseFunctions.get(currentToken.getType());
        AstNode leftExpression = prefix.get();
        while (!peekTokenIs(TokenType.SEMICOLON) && pte.ordinal() < peekPrecedence().ordinal()) {
            if (!infixParseFunctions.containsKey(peekToken.getType())) return leftExpression;
            Function<AstNode, AstNode> infix = infixParseFunctions.get(peekToken.getType());
            nextToken();
            leftExpression = infix.apply(leftExpression);
        }
        return leftExpression;
    }

    private AstNode parseInfixExpression(AstNode left) {
        AstInfixExpression expr = new AstInfixExpression();
        expr.token = getCurrentToken();
        expr.operator = getCurrentToken().literal;
        expr.left = left;
        PrecedenceTypeEnum precedence = currentPrecedence();
        nextToken();
        expr.right = parseExpression(precedence);
        return expr;
    }

    private AstNode parseIntegerLiteral() {
        String value = getCurrentToken().literal;
        return new AstInteger(getCurrentToken(), Integer.parseInt(value));
    }

    private PrecedenceTypeEnum currentPrecedence() {
        return precedences.get(currentToken.type);
    }

    private PrecedenceTypeEnum peekPrecedence() {
        if (precedences.containsKey(peekToken.type)) return precedences.get(peekToken.type);
        return PrecedenceTypeEnum.LOWEST;
    }

    private Token getCurrentToken() {
        return currentToken;
    }

    private boolean peekTokenIs(TokenType tokenType) {
        return peekToken.type == tokenType;
    }

    private void nextToken() {
        currentToken = peekToken;
        peekToken = lexer.nextToken();
    }

    public void load(String source) {
        this.source = source;
        lexer.load(source);
        nextToken();
        nextToken();
    }
}

File: cpp/repl/basic2/parser.cpp
#include "parser.h"

Parser::Parser() {
    prefixParseFns.insert(std::pair(TT_INT, &Parser::parseIntegerExpression));
    prefixParseFns.insert(std::pair(TT_LPAREN, &Parser::parseGroupedExpression));
    infixParseFns.insert(std::pair(TT_PLUS, &Parser::parseInfixExpression));
    infixParseFns.insert(std::pair(TT_MINUS, &Parser::parseInfixExpression));
    infixParseFns.insert(std::pair(TT_MULTIPLY, &Parser::parseInfixExpression));
    infixParseFns.insert(std::pair(TT_DIVIDE, &Parser::parseInfixExpression));
}

void Parser::load(std::string s) {
    source = s;
    lexer.load(source);
    nextToken();
    nextToken();
}

void Parser::nextToken() {
    current_token = peek_token;
    peek_token = lexer.nextToken();
}

Token Parser::currentToken() {
    return current_token;
}

Token Parser::peekToken() {
    return peek_token;
}

bool Parser::currentTokenIs(TokenTypeEnum tte) {
    return current_token.type == tte;
}

bool Parser::peekTokenIs(TokenTypeEnum tte) {
    return peek_token.type == tte;
}

bool Parser::expectPeek(TokenTypeEnum tte) {
    if (peekTokenIs(tte)) {
        nextToken();
        return true;
    } else {
        return false;
    }
}

PrecedenceTypeEnum Parser::currentPrecedence() {
    if (precedences.find(current_token.type) != precedences.end()) {
        return precedences[current_token.type];
    }
    return PT_LOWEST;
}

PrecedenceTypeEnum Parser::peekPrecedence() {
    if (precedences.find(peek_token.type) != precedences.end()) {
        return precedences[peek_token.type];
    }
    return PT_LOWEST;
}

void Parser::parseProgram(ASTProgram& program) {
    while(current_token.type != TT_EOF) {
        ASTNode* node = parseStatement();
        program.statements.push_back(node);
        nextToken();
    }
}

ASTNode* Parser::parseStatement() {
    // check token type for TT_LET, TT_RETURN, TT_COMMENT, etc..., if none of those then i have an expression that needs to be evaluated
    return parseExpressionStatement();
}

ASTNode* Parser::parseExpressionStatement() {
    ASTExpressionStatement *stmt = new ASTExpressionStatement(current_token);
    stmt->expression = parseExpression(PT_LOWEST);
    return stmt;
}

ASTNode* Parser::parseIntegerExpression() {
    ASTInteger *stmt = new ASTInteger(current_token);
    stmt->value = stoi(current_token.tokenLiteral);
    return stmt;
}

ASTNode* Parser::parseGroupedExpression() {
    nextToken();
    ASTNode* expr = parseExpression(PT_LOWEST);
    if (!expectPeek(TT_RPAREN)) {
        //free_Ad_AST_Node_memory(expr); // TODO: yeah, cleanup memory
        return NULL;
    }
    return expr;
}

ASTNode* Parser::parsePrefixExpression() {
    return NULL;
}

ASTNode* Parser::parseInfixExpression(ASTNode* left) {
    ASTInfixExpression *stmt = new ASTInfixExpression(current_token);
    stmt->_operator = current_token.tokenLiteral;
    stmt->left = left;
    PrecedenceTypeEnum current_precedence = currentPrecedence();
    nextToken();
    stmt->right = parseExpression(current_precedence);
    return stmt;
}

ASTNode* Parser::parseExpression(PrecedenceTypeEnum precedence) {
    if (prefixParseFns.find(current_token.type) == prefixParseFns.end()) {
        return NULL;
    }
    PrefixCallback prefix = prefixParseFns[current_token.type];

    ASTNode* leftExp = (this->*prefix)();

    while(!peekTokenIs(TT_SEMICOLON) && (precedence < peekPrecedence())) {
        if (infixParseFns.find(peek_token.type) == infixParseFns.end()) {
            return leftExp;
        }
        InfixCallback infix = infixParseFns[peek_token.type];
        nextToken();
        leftExp = (this->*infix)(leftExp);
    }
    return leftExp;
}